pkg/windows/service/service.go

```diff
@@ -31,20 +33,89 @@ import (
 )
 
 type handler struct {
-	tosvc   chan bool
-	fromsvc chan error
+	tosvc              chan bool
+	fromsvc            chan error
+	acceptPreshutdown  bool
+	preshutdownHandler PreshutdownHandler
 }
 
+type PreshutdownHandler interface {
+	ProcessShutdownEvent() error
+}
+
+// SERVICE_PRESHUTDOWN_INFO structure
+type SERVICE_PRESHUTDOWN_INFO struct {
+	PreshutdownTimeout uint32 // The time-out value, in milliseconds.
+}
+
+func QueryPreShutdownInfo(h windows.Handle) (*SERVICE_PRESHUTDOWN_INFO, error) {
+	// Query the SERVICE_CONFIG_PRESHUTDOWN_INFO
+	n := uint32(1024)
+	b := make([]byte, n)
+	for {
+		err := windows.QueryServiceConfig2(h, windows.SERVICE_CONFIG_PRESHUTDOWN_INFO, &b[0], n, &n)
+		if err == nil {
+			break
+		}
+		if err.(syscall.Errno) != syscall.ERROR_INSUFFICIENT_BUFFER {
+			return nil, err
+		}
+		if n <= uint32(len(b)) {
+			return nil, err
+		}
+
```

Comment:
Are you sure that other error cases don't need to be handled?

Reply:
it should be good enough, an example code is here: https://go.googlesource.com/sys/+/master/windows/svc/mgr/config.go#71